### ISSUE TITLE
TF addons Seq2Seq with Attention OOM when mixed_precision enable

### ISSUE THREAD

**Issue by dathudeptrai** [tensorflow/addons#2106]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): 18.04
- TensorFlow version and how it was installed (source or binary): 2.3.0
- TensorFlow-Addons version and how it was installed (source or binary): 0.10/0.11
- Python version: 3.7
- Is GPU used? (yes/no): Yes

**Describe the bug**

THe model OOM when mixed precision enable, without enable mixed precision, the model training fine.

**Code to reproduce the issue**

Here is my notebook to reproduce this bug. You can set `IS_MIXED_PRECISION` = True or False to check by your self. The model i implemented here is Tacotron-2. Note that the way i enable mixed precision works for other model in my framework except Tacotron-2.

[seq2seq.zip](https://github.com/tensorflow/addons/files/5102581/seq2seq.zip)

Provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Other info / logs**

Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

Hi, i'm creator of TensorFlowTTS (https://github.com/TensorSpeech/TensorFlowTTS). I'm trying to apply mixed precision for Tacotron-2 which implement by tensorflow_addons.seq2seq. Because the model training too slow so i want to use mixed_precision to speed-up the training progress but i failed. Hope you guys can help me fix this bug and that will help improve my user experience (also TF user experience :D.). Thanks a lot.

**Comment by bhack** [comment]
Could you open a very small PR extending seq2seq tests with a mixed precision failing case?

**Comment by dathudeptrai** [comment]
@bhack how can i do that?. Btw, the code is not small, it is a bit long because it is full pipeline end2end

**Comment by bhack** [comment]
What help us, if you can, It Is just to riproduce the same issue but with the minimal code snippet. 
When you have this It Is like a PR extending tests.

**Comment by dathudeptrai** [comment]
@bhack ok i will try to make a pull request to extend the test, meanwhile can you take a look ?

**Comment by bhack** [comment]
Could you share a colab?

**Comment by bhack** [comment]
Have you seen this warning:
`/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/indexed_slices.py:432: UserWarning: Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory.
  "Converting sparse IndexedSlices to a dense Tensor of unknown shape. `

**Comment by dathudeptrai** [comment]
> Have you seen this warning:
> `/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/indexed_slices.py:432: UserWarning: Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory. "Converting sparse IndexedSlices to a dense Tensor of unknown shape. `

yes :D. it's related to embedding layers ?

**Comment by bhack** [comment]
Yes it is related to the embedding layer. Check https://stackoverflow.com/questions/35892412/tensorflow-dense-gradient-explanation

**Comment by dathudeptrai** [comment]
> Yes it is related to the embedding layer. Check https://stackoverflow.com/questions/35892412/tensorflow-dense-gradient-explanation

thanks. But it's not the reason, i just remove embedding layers and replace it by fake tensor but still OOM

**Comment by bhack** [comment]
Can you share the new colab?

**Comment by dathudeptrai** [comment]
[seq2seq_new.zip](https://github.com/tensorflow/addons/files/5103468/seq2seq_new.zip)
here, i just random embedding layer output rather than use tf.keras.layers.Embedding(...)(input)

**Comment by bhack** [comment]
At what iteration is failing? Cause it is running fine on colab

**Comment by dathudeptrai** [comment]
> At what iteration is failing? Cause it is running fine on colab

because the colab can't apply mixed precision. The colab's GPU is old :D. You can check a log. Colab didn't enable mixed precision

**Comment by bhack** [comment]
It is only skipping auto_mixed_grap_optimizer but running mixed precision is supported as I see in the official tutorial:
(search colab) https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/guide/mixed_precision.ipynb

**Comment by dathudeptrai** [comment]
> It is only skipping auto_mixed_grap_optimizer but running mixed precision is supported as I see in the official tutorial:
> (search colab) https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/guide/mixed_precision.ipynb

Colab has 3 type GPU :)). Let use this code to check if ur colab support Mixed precision or not:

![Screenshot from 2020-08-20 21-10-05](https://user-images.githubusercontent.com/43868663/90782925-a0d99600-e329-11ea-8b11-a14b7302fb6b.png)

**Comment by bhack** [comment]
As you see in the tutorial:
> Note: If running this guide in Google Colab, the GPU runtime typically has a P100 connected. The P100 has compute capability 6.0 and is not expected to show a significant speedup.

So it is only about speedup.. But if it is also not running some other internal pass to reproduce your OOM error we need to wait for another maintainer with a local GPU for `float16`. 
Currently I have not this machine under my hands and I can only verify this with GPU in CI if you submit an isolated test/PR.

**Comment by dathudeptrai** [comment]
@guillaumekln Hi, i know you are a creator of openmt-tf and have many experience in seq2seq + mixed precision, can you help me ?. This kind of model play an important role in my framework (largest Tensorflow-2 Text to speech framework) so i just want to improve user experience. 

@bhack i will try my best to pr for testing mixed precision even it's hard i think.

**Comment by machineko** [comment]
I can confirm have same problem with OOM on my local rtx 2080TI while fp32 works fine.

**Comment by dathudeptrai** [comment]
@guillaumekln just remind :D

**Comment by guillaumekln** [comment]
I reproduced the issue but I'm not sure this is directly related to Addons. Auto mixed precision inserts a lot of cast ops (including on weights!) which can increase memory usage, especially in a long decoding loop such as the one in your example. I wonder if the converted weights are duplicated in each decoding step.

> because the colab can't apply mixed precision. The colab's GPU is old :D. You can check a log. Colab didn't enable mixed precision. it failed at the beginning, at iteration 0.

You can force the mixed precision graph rewrite by setting the following environment variable: 

```
TF_AUTO_MIXED_PRECISION_GRAPH_REWRITE_IGNORE_PERFORMANCE=1
```

**Comment by dathudeptrai** [comment]
@guillaumekln so i need report it to tensorflow repo ? or you are trying to solve this problem ?

**Comment by guillaumekln** [comment]
I don't see anything to fix in Addons, so you may try to raise this to the TensorFlow team.

**Comment by bhack** [comment]
Yes, please open a new ticket in the Tensorflow repo referencing this one.

**Comment by dathudeptrai** [comment]
thank :D i will try to report in tensorflow repo even the probability to get an answer is very low :))

**Comment by bhack** [comment]
@dathudeptrai We will upvote your ticket :wink:
